presto-hive/src/main/java/com/facebook/presto/hive/HiveLocationService.java

```diff
@@ -59,27 +64,43 @@ public LocationHandle forNewTable(SemiTransactionalHiveMetastore metastore, Conn
         if (pathExists(context, hdfsEnvironment, targetPath)) {
             throw new PrestoException(HIVE_PATH_ALREADY_EXISTS, format("Target directory for table '%s.%s' already exists: %s", schemaName, tableName, targetPath));
         }
-        return createLocationHandle(context, session, targetPath, false);
+        return createLocationHandle(context, session, targetPath, NEW);
     }
 
     @Override
     public LocationHandle forExistingTable(SemiTransactionalHiveMetastore metastore, ConnectorSession session, Table table)
     {
         HdfsContext context = new HdfsContext(session, table.getDatabaseName(), table.getTableName());
         Path targetPath = new Path(table.getStorage().getLocation());
-        return createLocationHandle(context, session, targetPath, true);
+        return createLocationHandle(context, session, targetPath, EXISTING);
     }
 
-    private LocationHandle createLocationHandle(HdfsContext context, ConnectorSession session, Path targetPath, boolean isExistingTable)
+    @Override
+    public LocationHandle forTemporaryTable(SemiTransactionalHiveMetastore metastore, ConnectorSession session, Table table)
+    {
+        String schemaName = table.getDatabaseName();
+        String tableName = table.getTableName();
+        HdfsContext context = new HdfsContext(session, schemaName, tableName);
+        Path targetPath = getTableDefaultLocation(context, metastore, hdfsEnvironment, schemaName, tableName);
+        // Cleanup target directory if exist
+        // Temporary table location is based on the table name that is globally unique
+        // This situation can happen if the insert into temporary table has been restarted
+        if (pathExists(context, hdfsEnvironment, targetPath)) {
+            deleteDirectory(context, hdfsEnvironment, targetPath);
```

Comment:
per offline discussion, just delete it would still be dangerous since some worker (after failure) might still try to write to the old directory. We should just fail for now :) .

In that case, you might want to revert the change to `testCreateBucketedTemporaryTable` that tries to write twice :)

Reply:
I think I'm just going to generate a new path for temporary table every time